drivers/flash/flash_andes_qspi_xip.c

```diff
@@ -0,0 +1,499 @@
+{
+	/* Disable prefetching to avoid any flash reads by hardware while working on flash. */
+	csr_clear(NDS_MCACHE_CTL, MCACHE_CTL_DPREF_EN);
+	csr_clear(NDS_MCACHE_CTL, MCACHE_CTL_IPREF_EN);
+
+	/* Finish all potetnial pending accesses to flash. */
+	__asm__ volatile("fence.i" ::: "memory");
+}
+
+static __ramfunc void cleanup_after_flashing(const struct device *dev, off_t addr, size_t size)
+{
+	const struct flash_andes_qspi_xip_config *config = dev->config;
+
+	/* Make sure all flash operations are finished. */
+	__asm__ volatile("fence.i" ::: "memory");
+
+	/* Invalidate modyfied flash memory. */
+	if (IS_ENABLED(CONFIG_CACHE_MANAGEMENT)) {
+		cache_data_invd_range((void *)(addr + config->mapped_base), size);
+		cache_instr_invd_range((void *)(addr + config->mapped_base), size);
+	}
+
+	/* Restore prefetching */
+	csr_set(NDS_MCACHE_CTL, MCACHE_CTL_IPREF_EN);
+	csr_set(NDS_MCACHE_CTL, MCACHE_CTL_DPREF_EN);
+}
+
+static int flash_andes_qspi_xip_read(const struct device *dev, off_t addr, void *dest, size_t size)
+{
+	const struct flash_andes_qspi_xip_config *config = dev->config;
+
+	if (size == 0) {
+		return 0;
+	}
+
+	/* Check read boundaries */
+	if ((addr < 0 || addr >= config->flash_size) || ((config->flash_size - addr) < size)) {
+		return -EINVAL;
+	}
+
+	/* Use memory-mapped mechanism for reading. */
+	memcpy(dest, (void *)(config->mapped_base + addr), size);
+
+	return 0;
+}
+
+static __ramfunc int do_write(const struct device *dev, off_t addr, const void *src, size_t size)
+{
+	int ret;
+	size_t to_write;
+	off_t addr_curr = addr;
+	size_t size_remainig = size;
+
+	prepare_for_flashing();
+
+	while (size_remainig > 0) {
+		ret = write_protection_set(dev, false);
+		if (ret != 0) {
+			break;
+		}
+
+		/* Get the adequate size to send*/
+		to_write = MIN(PAGE_SIZE - (addr_curr % PAGE_SIZE), size_remainig);
+		ret = flash_andes_qspi_xip_cmd_addr_write(dev, SPI_NOR_CMD_PP_1_1_4, addr_curr, src,
+							  to_write);
+		flash_andes_qspi_xip_wait_until_ready(dev);
+
+		if (ret != 0) {
+			break;
+		}
+
+		src = (const uint8_t *)src + to_write;
+		size_remainig -= to_write;
+		addr_curr += to_write;
+	}
+
+	/* Make sure write protection is enabled at the end of write. */
+	int ret2 = write_protection_set(dev, true);
+
+	if (!ret) {
+		ret = ret2;
+	}
+
+	cleanup_after_flashing(dev, addr, size);
+
+	return ret;
+}
+
+static int flash_andes_qspi_xip_write(const struct device *dev, off_t addr, const void *src,
+				      size_t size)
+{
+	const struct flash_andes_qspi_xip_config *config = dev->config;
+	int ret = 0;
+	unsigned int key;
+
+	if (size == 0) {
+		return 0;
+	}
+
+	/* Check write boundaries */
+	if ((addr < 0 || addr >= config->flash_size) || ((config->flash_size - addr) < size)) {
+		return -EINVAL;
+	}
+
+	/*
+	 * Synchronous mechanisms like semaphores are not needed,
+	 * because interrupts are always locked and there are no reschedule points.
+	 */
+	key = irq_lock();
+
+	ret = do_write(dev, addr, src, size);
```

Comment:
On the Andes core, it is recommended to disable mmisc_ctl.BRPE before jumping to __ramfunc to avoid potential issues.
This is because the branch prediction unit on Andes cores includes a Return Address Stack (RAS) feature, which may cause the CPU to prefetch instructions at the return address (RA) that in flash memory(in this case) while executing on RAM.
And re-enable it after returning.

Reply:
Can use
`csr_clear(NDS_MMISC_CTL, BIT(3));`
and
`csr_set(NDS_MMISC_CTL, BIT(3));`

Reply:
Disabling it before jumping to __ramfunc is to further ensure that the following ra (in flash) is not pushed onto the Return Address Stack (RAS).

Reply:
Thanks for the explanation, done.

Reply:
`prepare_for_ramfunc `and `cleanup_after_ramfunc `should not be the __ramfunc function.
They are used to disable/enable mmisc_ctl.BRPE before/after the __ramfunc, therefore they should not be executed within __ramfunc.

The rest of the changes are all LGTM!

Reply:
I agree. Done.